In [1]:

import pandas as pd
import numpy as np
import re
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import scipy.stats as stats


In [2]:
df_final_demo = r'https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_demo.txt'
df_final_web_data_pt1 = r'https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_web_data_pt_1.txt'
df_final_web_data_pt2 = r'https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_web_data_pt_2.txt'
df_final_experiment_client = r'https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_experiment_clients.txt'



In [ ]:
# Intenta leer el archivo
try:
    df_final_demo = pd.read_csv(df_final_demo, sep=',')  
    df_final_web_data_pt1= pd.read_csv(df_final_web_data_pt1, sep=',')
    df_final_web_data_pt2= pd.read_csv(df_final_web_data_pt2, sep=',')
    df_final_experiment_client= pd.read_csv(df_final_experiment_client, sep=',')
except Exception as e:
    print(f"Error al cargar el archivo: {e}")

Error al cargar el archivo: argument of type 'method' is not iterable


In [48]:
df_final_demo.columns = df_final_demo.columns.str.lower()
df_final_web_data_pt1.columns = df_final_web_data_pt1.columns.str.lower()
df_final_web_data_pt2.columns = df_final_web_data_pt2.columns.str.lower()
df_final_experiment_client.columns = df_final_experiment_client.columns.str.lower()


In [49]:
final_demo = df_final_demo
final_demo

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0         836976             6.0              73.0      60.5     U        2.0   
1        2304905             7.0              94.0      58.0     U        2.0   
2        1439522             5.0              64.0      32.0     U        2.0   
3        1562045            16.0             198.0      49.0     M        2.0   
4        5126305            12.0             145.0      33.0     F        2.0   
...          ...             ...               ...       ...   ...        ...   
70604    7993686             4.0              56.0      38.5     U        3.0   
70605    8981690            12.0             148.0      31.0     M        2.0   
70606     333913            16.0             198.0      61.5     F        2.0   
70607    1573142            21.0             255.0      68.0     M        3.0   
70608    5602139            21.0             254.0      59.5     F        3.0   

              bal  calls_6_mnth  logons_6_mnth  
0        45105.30           6.0            9.0  
1       110860.30           6.0            9.0  
2        52467.79           6.0            9.0  
3        67454.65           3.0            6.0  
4       103671.75           0.0            3.0  
...           ...           ...            ...  
70604  1411062.68           5.0            5.0  
70605   101867.07           6.0            6.0  
70606    40745.00           3.0            3.0  
70607   475114.69           4.0            4.0  
70608   157498.73           7.0            7.0  

[70609 rows x 9 columns]

In [50]:
final_demo = pd.merge(final_demo, df_final_experiment_client, on='client_id')
final_demo

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0         836976             6.0              73.0      60.5     U        2.0   
1        2304905             7.0              94.0      58.0     U        2.0   
2        1439522             5.0              64.0      32.0     U        2.0   
3        1562045            16.0             198.0      49.0     M        2.0   
4        5126305            12.0             145.0      33.0     F        2.0   
...          ...             ...               ...       ...   ...        ...   
70604    7993686             4.0              56.0      38.5     U        3.0   
70605    8981690            12.0             148.0      31.0     M        2.0   
70606     333913            16.0             198.0      61.5     F        2.0   
70607    1573142            21.0             255.0      68.0     M        3.0   
70608    5602139            21.0             254.0      59.5     F        3.0   

              bal  calls_6_mnth  logons_6_mnth variation  
0        45105.30           6.0            9.0      Test  
1       110860.30           6.0            9.0   Control  
2        52467.79           6.0            9.0      Test  
3        67454.65           3.0            6.0      Test  
4       103671.75           0.0            3.0   Control  
...           ...           ...            ...       ...  
70604  1411062.68           5.0            5.0       NaN  
70605   101867.07           6.0            6.0       NaN  
70606    40745.00           3.0            3.0       NaN  
70607   475114.69           4.0            4.0       NaN  
70608   157498.73           7.0            7.0       NaN  

[70609 rows x 10 columns]

In [51]:
final_demo = final_demo.dropna()
final_demo

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0         836976             6.0              73.0      60.5     U        2.0   
1        2304905             7.0              94.0      58.0     U        2.0   
2        1439522             5.0              64.0      32.0     U        2.0   
3        1562045            16.0             198.0      49.0     M        2.0   
4        5126305            12.0             145.0      33.0     F        2.0   
...          ...             ...               ...       ...   ...        ...   
50495    1780858            21.0             262.0      68.5     M        3.0   
50496    6967120            21.0             260.0      68.5     M        3.0   
50497    5826160            20.0             249.0      56.5     F        2.0   
50498    8739285            19.0             229.0      69.5     F        2.0   
50499    7468138            18.0             222.0      61.0     F        3.0   

              bal  calls_6_mnth  logons_6_mnth variation  
0        45105.30           6.0            9.0      Test  
1       110860.30           6.0            9.0   Control  
2        52467.79           6.0            9.0      Test  
3        67454.65           3.0            6.0      Test  
4       103671.75           0.0            3.0   Control  
...           ...           ...            ...       ...  
50495   372100.59           6.0            9.0      Test  
50496  4279873.38           6.0            9.0   Control  
50497    44837.16           2.0            5.0      Test  
50498    44994.24           1.0            4.0      Test  
50499   209278.15           0.0            3.0      Test  

[50487 rows x 10 columns]

In [53]:
df_final_web_data_pt1

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
343136    2443347  465784886_73090545671  136329900_10529659391_316129   
343137    2443347  465784886_73090545671  136329900_10529659391_316129   
343138    2443347  465784886_73090545671  136329900_10529659391_316129   
343139    2443347  465784886_73090545671  136329900_10529659391_316129   
343140    2443347  465784886_73090545671  136329900_10529659391_316129   

       process_step            date_time  
0            step_3  2017-04-17 15:27:07  
1            step_2  2017-04-17 15:26:51  
2            step_3  2017-04-17 15:19:22  
3            step_2  2017-04-17 15:19:13  
4            step_3  2017-04-17 15:18:04  
...             ...                  ...  
343136      confirm  2017-03-31 15:15:46  
343137       step_3  2017-03-31 15:14:53  
343138       step_2  2017-03-31 15:12:08  
343139       step_1  2017-03-31 15:11:37  
343140        start  2017-03-31 15:10:47  

[343141 rows x 5 columns]

In [54]:
df_final_web_data_pt2

client_id             visitor_id                      visit_id  \
0          763412  601952081_10457207388  397475557_40440946728_419634   
1         6019349  442094451_91531546617  154620534_35331068705_522317   
2         6019349  442094451_91531546617  154620534_35331068705_522317   
3         6019349  442094451_91531546617  154620534_35331068705_522317   
4         6019349  442094451_91531546617  154620534_35331068705_522317   
...           ...                    ...                           ...   
412259    9668240   388766751_9038881013   922267647_3096648104_968866   
412260    9668240   388766751_9038881013   922267647_3096648104_968866   
412261    9668240   388766751_9038881013   922267647_3096648104_968866   
412262    9668240   388766751_9038881013   922267647_3096648104_968866   
412263     674799  947159805_81558194550   86152093_47511127657_716022   

       process_step            date_time  
0           confirm  2017-06-06 08:56:00  
1           confirm  2017-06-01 11:59:27  
2            step_3  2017-06-01 11:58:48  
3            step_2  2017-06-01 11:58:08  
4            step_1  2017-06-01 11:57:58  
...             ...                  ...  
412259        start  2017-05-24 18:46:10  
412260        start  2017-05-24 18:45:29  
412261       step_1  2017-05-24 18:44:51  
412262        start  2017-05-24 18:44:34  
412263        start  2017-06-03 12:17:09  

[412264 rows x 5 columns]

In [55]:
final_web = pd.concat([df_final_web_data_pt1, df_final_web_data_pt2], ignore_index=True)
final_web

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
755400    9668240   388766751_9038881013   922267647_3096648104_968866   
755401    9668240   388766751_9038881013   922267647_3096648104_968866   
755402    9668240   388766751_9038881013   922267647_3096648104_968866   
755403    9668240   388766751_9038881013   922267647_3096648104_968866   
755404     674799  947159805_81558194550   86152093_47511127657_716022   

       process_step            date_time  
0            step_3  2017-04-17 15:27:07  
1            step_2  2017-04-17 15:26:51  
2            step_3  2017-04-17 15:19:22  
3            step_2  2017-04-17 15:19:13  
4            step_3  2017-04-17 15:18:04  
...             ...                  ...  
755400        start  2017-05-24 18:46:10  
755401        start  2017-05-24 18:45:29  
755402       step_1  2017-05-24 18:44:51  
755403        start  2017-05-24 18:44:34  
755404        start  2017-06-03 12:17:09  

[755405 rows x 5 columns]

In [56]:
final_web.isna()

client_id  visitor_id  visit_id  process_step  date_time
0           False       False     False         False      False
1           False       False     False         False      False
2           False       False     False         False      False
3           False       False     False         False      False
4           False       False     False         False      False
...           ...         ...       ...           ...        ...
755400      False       False     False         False      False
755401      False       False     False         False      False
755402      False       False     False         False      False
755403      False       False     False         False      False
755404      False       False     False         False      False

[755405 rows x 5 columns]

In [57]:
print(final_web.isnull().sum())


client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64


In [70]:
tabla_analisis = pd.merge(final_demo, final_web, on='client_id')

In [71]:
tabla_analisis

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  \
0          836976             6.0              73.0      60.5     U   
1          836976             6.0              73.0      60.5     U   
2          836976             6.0              73.0      60.5     U   
3          836976             6.0              73.0      60.5     U   
4          836976             6.0              73.0      60.5     U   
...           ...             ...               ...       ...   ...   
321190    7468138            18.0             222.0      61.0     F   
321191    7468138            18.0             222.0      61.0     F   
321192    7468138            18.0             222.0      61.0     F   
321193    7468138            18.0             222.0      61.0     F   
321194    7468138            18.0             222.0      61.0     F   

        num_accts        bal  calls_6_mnth  logons_6_mnth variation  \
0             2.0   45105.30           6.0            9.0      Test   
1             2.0   45105.30           6.0            9.0      Test   
2             2.0   45105.30           6.0            9.0      Test   
3             2.0   45105.30           6.0            9.0      Test   
4             2.0   45105.30           6.0            9.0      Test   
...           ...        ...           ...            ...       ...   
321190        3.0  209278.15           0.0            3.0      Test   
321191        3.0  209278.15           0.0            3.0      Test   
321192        3.0  209278.15           0.0            3.0      Test   
321193        3.0  209278.15           0.0            3.0      Test   
321194        3.0  209278.15           0.0            3.0      Test   

                   visitor_id                      visit_id process_step  \
0        427070339_1413275162   228976764_46825473280_96584      confirm   
1        427070339_1413275162   228976764_46825473280_96584      confirm   
2        427070339_1413275162   228976764_46825473280_96584      confirm   
3        427070339_1413275162   228976764_46825473280_96584       step_3   
4        427070339_1413275162   228976764_46825473280_96584       step_2   
...                       ...                           ...          ...   
321190  317123960_46820919455  769876461_30381166055_830233       step_2   
321191  317123960_46820919455  769876461_30381166055_830233       step_1   
321192  317123960_46820919455  769876461_30381166055_830233        start   
321193  317123960_46820919455  769876461_30381166055_830233        start   
321194  317123960_46820919455  769876461_30381166055_830233        start   

                  date_time  
0       2017-04-02 11:51:13  
1       2017-04-02 11:47:50  
2       2017-04-02 11:46:45  
3       2017-04-02 11:23:08  
4       2017-04-02 11:22:24  
...                     ...  
321190  2017-03-30 23:59:15  
321191  2017-03-30 23:58:51  
321192  2017-03-30 23:58:40  
321193  2017-03-30 23:55:11  
321194  2017-03-31 00:16:12  

[321195 rows x 14 columns]

In [72]:
tabla_analisis.head()

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0     836976             6.0              73.0      60.5     U        2.0   
1     836976             6.0              73.0      60.5     U        2.0   
2     836976             6.0              73.0      60.5     U        2.0   
3     836976             6.0              73.0      60.5     U        2.0   
4     836976             6.0              73.0      60.5     U        2.0   

       bal  calls_6_mnth  logons_6_mnth variation            visitor_id  \
0  45105.3           6.0            9.0      Test  427070339_1413275162   
1  45105.3           6.0            9.0      Test  427070339_1413275162   
2  45105.3           6.0            9.0      Test  427070339_1413275162   
3  45105.3           6.0            9.0      Test  427070339_1413275162   
4  45105.3           6.0            9.0      Test  427070339_1413275162   

                      visit_id process_step            date_time  
0  228976764_46825473280_96584      confirm  2017-04-02 11:51:13  
1  228976764_46825473280_96584      confirm  2017-04-02 11:47:50  
2  228976764_46825473280_96584      confirm  2017-04-02 11:46:45  
3  228976764_46825473280_96584       step_3  2017-04-02 11:23:08  
4  228976764_46825473280_96584       step_2  2017-04-02 11:22:24

In [73]:
tabla_analisis["process_step"].unique()

array(['confirm', 'step_3', 'step_2', 'step_1', 'start'], dtype=object)

In [ ]:
tabla_test = tabla_analisis.set_index(['client_id', 'process_step'])
tabla_test.tail(10)

clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  \
client_id process_step                                                     
8739285   step_3                  19.0             229.0      69.5     F   
          step_2                  19.0             229.0      69.5     F   
          step_1                  19.0             229.0      69.5     F   
          start                   19.0             229.0      69.5     F   
7468138   step_3                  18.0             222.0      61.0     F   
          step_2                  18.0             222.0      61.0     F   
          step_1                  18.0             222.0      61.0     F   
          start                   18.0             222.0      61.0     F   
          start                   18.0             222.0      61.0     F   
          start                   18.0             222.0      61.0     F   

                        num_accts        bal  calls_6_mnth  logons_6_mnth  \
client_id process_step                                                      
8739285   step_3              2.0   44994.24           1.0            4.0   
          step_2              2.0   44994.24           1.0            4.0   
          step_1              2.0   44994.24           1.0            4.0   
          start               2.0   44994.24           1.0            4.0   
7468138   step_3              3.0  209278.15           0.0            3.0   
          step_2              3.0  209278.15           0.0            3.0   
          step_1              3.0  209278.15           0.0            3.0   
          start               3.0  209278.15           0.0            3.0   
          start               3.0  209278.15           0.0            3.0   
          start               3.0  209278.15           0.0            3.0   

                       variation             visitor_id  \
client_id process_step                                    
8739285   step_3            Test    8756552_54841419737   
          step_2            Test    8756552_54841419737   
          step_1            Test    8756552_54841419737   
          start             Test    8756552_54841419737   
7468138   step_3            Test  317123960_46820919455   
          step_2            Test  317123960_46820919455   
          step_1            Test  317123960_46820919455   
          start             Test  317123960_46820919455   
          start             Test  317123960_46820919455   
          start             Test  317123960_46820919455   

                                            visit_id            date_time  
client_id process_step                                                     
8739285   step_3        101846948_96925910026_299041  2017-04-12 14:58:39  
          step_2        101846948_96925910026_299041  2017-04-12 14:58:16  
          step_1        101846948_96925910026_299041  2017-04-12 14:50:44  
          start         101846948_96925910026_299041  2017-04-12 14:49:47  
7468138   step_3        769876461_30381166055_830233  2017-03-30 23:59:33  
          step_2        769876461_30381166055_830233  2017-03-30 23:59:15  
          step_1        769876461_30381166055_830233  2017-03-30 23:58:51  
          start         769876461_30381166055_830233  2017-03-30 23:58:40  
          start         769876461_30381166055_830233  2017-03-30 23:55:11  
          start         769876461_30381166055_830233  2017-03-31 00:16:12

In [ ]:
tabla_test

In [80]:
tabla_analisis_reducida = tabla_analisis[['client_id', 'process_step']]
tabla_analisis_reducida

client_id process_step
0          836976      confirm
1          836976      confirm
2          836976      confirm
3          836976       step_3
4          836976       step_2
...           ...          ...
321190    7468138       step_2
321191    7468138       step_1
321192    7468138        start
321193    7468138        start
321194    7468138        start

[321195 rows x 2 columns]

In [91]:
pivot_df = tabla_analisis.pivot_table(index='client_id', columns='process_step', aggfunc='size', fill_value=0)
pivot_df

process_step  confirm  start  step_1  step_2  step_3
client_id                                           
555                 1      1       1       1       1
647                 1      1       1       1       1
934                 0      4       0       0       0
1028                0      1       5       2       1
1104                0      2       0       0       0
...               ...    ...     ...     ...     ...
9999150             0      2       1       0       0
9999400             1      1       1       1       1
9999626             0      1       1       0       0
9999729             1      4       3       2       1
9999832             0      1       1       0       0

[50487 rows x 5 columns]

In [102]:
clients_at_start = pivot_df[pivot_df['start'] > 0]
clients_at_start

process_step  confirm  start  step_1  step_2  step_3
client_id                                           
555                 1      1       1       1       1
647                 1      1       1       1       1
934                 0      4       0       0       0
1028                0      1       5       2       1
1104                0      2       0       0       0
...               ...    ...     ...     ...     ...
9999150             0      2       1       0       0
9999400             1      1       1       1       1
9999626             0      1       1       0       0
9999729             1      4       3       2       1
9999832             0      1       1       0       0

[50063 rows x 5 columns]

In [103]:
clients_at_step_1 = pivot_df[pivot_df['step_1'] > 0]
clients_at_step_1

process_step  confirm  start  step_1  step_2  step_3
client_id                                           
555                 1      1       1       1       1
647                 1      1       1       1       1
1028                0      1       5       2       1
1186                0      2       1       1       0
1195                1      1       1       1       1
...               ...    ...     ...     ...     ...
9999150             0      2       1       0       0
9999400             1      1       1       1       1
9999626             0      1       1       0       0
9999729             1      4       3       2       1
9999832             0      1       1       0       0

[44406 rows x 5 columns]

In [104]:
clients_at_step_2 = pivot_df[pivot_df['step_2'] > 0]
clients_at_step_2

process_step  confirm  start  step_1  step_2  step_3
client_id                                           
555                 1      1       1       1       1
647                 1      1       1       1       1
1028                0      1       5       2       1
1186                0      2       1       1       0
1195                1      1       1       1       1
...               ...    ...     ...     ...     ...
9998851             1      1       1       1       1
9998964             1      1       1       1       1
9999009             1      1       1       1       1
9999400             1      1       1       1       1
9999729             1      4       3       2       1

[40896 rows x 5 columns]

In [105]:
clients_at_step_3 = pivot_df[pivot_df['step_3'] > 0]
clients_at_step_3

process_step  confirm  start  step_1  step_2  step_3
client_id                                           
555                 1      1       1       1       1
647                 1      1       1       1       1
1028                0      1       5       2       1
1195                1      1       1       1       1
1197                1      1       1       2       2
...               ...    ...     ...     ...     ...
9998851             1      1       1       1       1
9998964             1      1       1       1       1
9999009             1      1       1       1       1
9999400             1      1       1       1       1
9999729             1      4       3       2       1

[38292 rows x 5 columns]

In [ ]:
# Filtrar las filas donde 'process_step' es igual a 'confirm'
clients_at_confirm = pivot_df[pivot_df['confirm'] > 0]
clients_at_confirm


process_step  confirm  start  step_1  step_2  step_3
client_id                                           
555                 1      1       1       1       1
647                 1      1       1       1       1
1195                1      1       1       1       1
1197                1      1       1       2       2
1336                3      1       1       1       1
...               ...    ...     ...     ...     ...
9998851             1      1       1       1       1
9998964             1      1       1       1       1
9999009             1      1       1       1       1
9999400             1      1       1       1       1
9999729             1      4       3       2       1

[34110 rows x 5 columns]

In [119]:
confirm = len(clients_at_confirm)
start = len(clients_at_start)
step1 = len(clients_at_step_1)
step2 = len(clients_at_step_2)
step3 = len(clients_at_step_3)
total_usuarios = len(df_final_demo)
total_test = len(pivot_df)

In [120]:
total_usuarios, total_test, start ,step1, step2 ,step3 , confirm

(70609, 50487, 50063, 44406, 40896, 38292, 34110)

In [121]:
d_index = {
    "name": ['total_usuarios', 'total_test', 'start' ,'step1', 'step2' ,'step3' , 'confirm'],
    "score": [total_usuarios, total_test, start ,step1, step2 ,step3 , confirm]
}

df_index = pd.DataFrame(d_index)
df_index

name  score
0  total_usuarios  70609
1      total_test  50487
2           start  50063
3           step1  44406
4           step2  40896
5           step3  38292
6         confirm  34110

TOTAL CLIENTES= 70,609
START = 50,063
STEP 1 = 44,406
STEP 2 = 40,896
STEP 3 = 38,292
CONFIRM = 34,110
% de caída de START a STEP 1: 11.37%
% de caída de STEP 1 a STEP 2: 7.97%
% de caída de STEP 2 a STEP 3: 6.37%
% de caída de STEP 3 a CONFIRM: 10.92%
% de clientes que han confirmado: 48.3%

In [ ]:
Hipotesis_n1 = [total_empresa, confirm, step1, step2, step3], axis=1)

TypeError: cannot concatenate object of type '<class 'int'>'; only Series and DataFrame objs are valid

In [87]:
pivot_df.sum()

process_step
confirm     43214
start      104041
step_1      68410
step_2      56855
step_3      48675
dtype: int64

In [62]:
tabla_analisis.describe()

client_id  clnt_tenure_yr  clnt_tenure_mnth       clnt_age  \
count  3.211950e+05   321195.000000     321195.000000  321195.000000   
mean   5.009766e+06       12.176528        152.134818      48.553511   
std    2.872653e+06        6.963181         83.210462      15.645588   
min    5.550000e+02        2.000000         33.000000      17.000000   
25%    2.514553e+06        6.000000         82.000000      34.500000   
50%    5.051667e+06       11.000000        138.000000      50.000000   
75%    7.468390e+06       16.000000        193.000000      61.000000   
max    9.999832e+06       55.000000        669.000000      96.000000   

           num_accts           bal   calls_6_mnth  logons_6_mnth  
count  321195.000000  3.211950e+05  321195.000000  321195.000000  
mean        2.262962  1.622142e+05       3.236834       6.275082  
std         0.541536  3.494776e+05       2.193690       2.179757  
min         1.000000  2.378944e+04       0.000000       3.000000  
25%         2.000000  4.105887e+04       1.000000       4.000000  
50%         2.000000  6.924195e+04       3.000000       6.000000  
75%         2.000000  1.514930e+05       6.000000       9.000000  
max         7.000000  1.632004e+07       6.000000       9.000000